src/HIE/Bios/Types.hs

```diff
@@ -44,21 +44,60 @@ data ActionName a
   | Other a
   deriving (Show, Eq, Ord, Functor)
 
-data CradleAction a = CradleAction {
-                        actionName    :: ActionName a
-                      -- ^ Name of the action.
-                      , runCradle     :: LoggingFunction -> FilePath -> IO (CradleLoadResult ComponentOptions)
-                      -- ^ Options to compile the given file with.
-                      , runGhcCmd :: [String] -> IO (CradleLoadResult String)
-                      -- ^ Executes the @ghc@ binary that is usually used to
-                      -- build the cradle. E.g. for a cabal cradle this should be
-                      -- equivalent to @cabal exec ghc -- args@
-                      }
+data CradleAction a = CradleAction
+  { actionName    :: ActionName a
+  -- ^ Name of the action.
+  , runCradle     :: LoggingFunction -> FilePath -> IO (CradleLoadResult LoadResult)
+  -- ^ Options to compile the given file with.
+  --
+  -- The given FilePath /must/ be part of 'LoadResult.loadResultComponent' if
+  -- the loading operation succeeds.
+  , runGhcCmd :: [String] -> IO (CradleLoadResult String)
+  -- ^ Executes the @ghc@ binary that is usually used to
+  -- build the cradle. E.g. for a cabal cradle this should be
+  -- equivalent to @cabal exec ghc -- args@
+  }
   deriving (Functor)
 
 instance Show a => Show (CradleAction a) where
   show CradleAction { actionName = name } = "CradleAction: " ++ show name
 
+type LoadResult = LoadResult' ComponentOptions
+
+-- | Record for expressing successful loading.
+-- Can express partial success.
+data LoadResult' a = LoadResult
+  { loadResultComponent :: Maybe a
+  -- ^ Component options for the FilePath that produced this 'LoadResult'.
+  -- See 'CradleAction.runCradle' for information on how to produce a 'LoadResult'.
+  --
+  -- This field can be 'Nothing' to indicate that loading partially failed/succeeded.
+  , loadResultDependencies :: [a]
+  -- ^ Direct or indirect dependencies from the component from above.
+  -- Indirect means that it is not required that 'ComponentOptions' in this list
+  -- are required dependencies of 'loadResultComponent'. It is specifically allowed
```

Comment:
The sentence makes no sense, what I mean is, that items here are either local dependencies in some form, or completely independent but can be loaded, too.
It might still be important to have some kind of order on them.